### PR TITLE
HADOOP-17079. Optimize UGI#getGroups by adding UGI#getGroupsSet.

### PR DESCRIPTION
**Pull request by xiaoyuyao**
## NOTICE

https://issues.apache.org/jira/browse/HADOOP-17079

UserGroupInformation.java adding getGroupsSet() that returns a Set instead of List. The callers of Groups#getGroups() in Hadoop are replaced with getGroupsSet() as we know Set#contains() will be
generally faster than List#contains expecially the user has large group memeberships. In some of our cusotmers' datalake use cases, user can typically have hundreds or thousands of group members. Changing from list to set make a big difference. 

GroupMappingServiceProvider.java add a new method that returns result as Set instead of List
Groups.java Change to maintain cached user -> groups mapping as a set instead of list.

### CODE REVIEW

**Comment by xiaoyuyao** [comment]
@jojochuang do I miss anything to trigger a Jenkins run for this PR? I also add a link to the patch in the JIRA based on the "Hadoop  How to Contribute document on PR". It does not seem to work either.

**Comment by Hexiaoqiao** [comment]
Hi @xiaoyuyao , it could need trigger Jenkins manually following this guide, https://cwiki.apache.org/confluence/display/HADOOP/GitHub+Integration#GitHubIntegration-HowtorunJenkinsprecommitjobforaPR(committers). I try to trigger it, please ref: https://builds.apache.org/view/H-L/view/Hadoop/job/hadoop-multibranch/view/change-requests/job/PR-2085/2/console

File: hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestRouterUserMappings.java [hunk 1]
```diff
@@ -111,6 +112,11 @@ public void cacheGroupsRefresh() throws IOException {
     @Override
     public void cacheGroupsAdd(List<String> groups) throws IOException {
     }
+
+    @Override
+    public Set<String> getGroupsSet(String user) throws IOException {
+      return null;
```
Comment (Hexiaoqiao) [inline hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestRouterUserMappings.java:118]: Do we need keep the same logic with getGroups. Not sure if this method will invoke by other unit test.
```
    @Override
    public List<String> getGroups(String user) throws IOException {
      LOG.info("Getting groups in MockUnixGroupsMapping");
      String g1 = user + (10 * i + 1);
      String g2 = user + (10 * i + 2);
      List<String> l = new ArrayList<String>(2);
      l.add(g1);
      l.add(g2);
      i++;
      return l;
    }
```

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSPermissionChecker.java [hunk 2]
```diff
@@ -549,7 +549,6 @@ private boolean hasPermission(INodeAttributes inode, FsAction access) {
    * - Default entries may be present, but they are ignored during enforcement.
    *
    * @param inode INodeAttributes accessed inode
-   * @param snapshotId int snapshot ID
```
Comment (Hexiaoqiao) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSPermissionChecker.java:552]: It seems that is not related to this changes.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/Groups.java [hunk 3]
```diff
@@ -345,28 +373,28 @@ public long read() {
...
       LOG.debug("GroupCacheLoader - reload (async).");
       if (!reloadGroupsInBackground) {
         return super.reload(key, oldValue);
       }
 
       backgroundRefreshQueued.incrementAndGet();
-      ListenableFuture<List<String>> listenableFuture =
-          executorService.submit(new Callable<List<String>>() {
+      ListenableFuture<Set<String>> listenableFuture =
+          executorService.submit(new Callable<Set<String>>() {
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/Groups.java:386]: Good catch. Fixed.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/JniBasedUnixGroupsMapping.java [hunk 4]
```diff
@@ -19,9 +19,9 @@
 package org.apache.hadoop.security;
 
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/JniBasedUnixGroupsMapping.java:22]: IntelliJ auto folded the imports. I will fix them in the next commit.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/NullGroupsMapping.java [hunk 5]
```diff
@@ -31,6 +33,19 @@
...
+   * Get all various group memberships of a given user.
+   * Returns EMPTY set in case of non-existing user
+   *
+   * @param user User's name
+   * @return set of group memberships of user
+   * @throws IOException
+   */
+  @Override
+  public Set<String> getGroupsSet(String user) throws IOException {
+    return null;
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/NullGroupsMapping.java:46]: Fixed.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/authorize/AccessControlList.java [hunk 6]
```diff
@@ -20,10 +20,7 @@
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
-import java.util.Collection;
-import java.util.HashSet;
-import java.util.LinkedList;
-import java.util.List;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/authorize/AccessControlList.java:23]: Fixed.

File: hadoop-hdfs-project/hadoop-hdfs-httpfs/src/test/java/org/apache/hadoop/lib/service/security/DummyGroupMapping.java [hunk 7]
```diff
@@ -47,4 +48,9 @@ public void cacheGroupsRefresh() throws IOException {
   @Override
   public void cacheGroupsAdd(List<String> groups) throws IOException {
   }
+
+  @Override
+  public Set<String> getGroupsSet(String user) throws IOException {
+    return null;
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-hdfs-project/hadoop-hdfs-httpfs/src/test/java/org/apache/hadoop/lib/service/security/DummyGroupMapping.java:54]: Fixed.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSPermissionChecker.java (hunk 2 above)
Comment (xiaoyuyao) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSPermissionChecker.java:552]: I just reverted this change.

File: hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestRouterUserMappings.java (hunk 1 above)
Comment (xiaoyuyao) [reply hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestRouterUserMappings.java:118]: Fixed with similar logic.

**Comment by xiaoyuyao** [comment]
Thanks @Hexiaoqiao  fro the review and positive results of similar changes in your deployment. I've  addressed the feedback in the new commit.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/Groups.java [hunk 8]
```diff
@@ -203,17 +202,47 @@ private IOException noGroupsForUser(String user) {
   /**
    * Get the group memberships of a given user.
    * If the user's group is not cached, this method may block.
+   * Note this method can be expensive as it involves Set->List conversion.
+   * For user with large group membership (i.e., > 1000 groups), we recommend
+   * using getGroupSet to avoid the conversion and fast membership look up via
+   * contains().
```
Comment (jojochuang) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/Groups.java:208]: Would it make sense to deprecate this API to push users to use the new API?

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/Groups.java [hunk 9]
```diff
@@ -203,17 +202,47 @@ private IOException noGroupsForUser(String user) {
...
+   */
+  private Set<String> getGroupInternal(final String user) throws IOException {
     // No need to lookup for groups of static users
-    Map<String, List<String>> staticUserToGroupsMap = staticMapRef.get();
+    Map<String, Set<String>> staticUserToGroupsMap = staticMapRef.get();
     if (staticUserToGroupsMap != null) {
-      List<String> staticMapping = staticUserToGroupsMap.get(user);
+      Set<String> staticMapping = staticUserToGroupsMap.get(user);
       if (staticMapping != null) {
-        return staticMapping;
```
Comment (xiaoyuyao) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/Groups.java:216]: No need to add Collections.unmodifiableSet here as the caller from the public API will do it.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 19s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 15 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  20m 14s |  trunk passed  |
| +1 :green_heart: |  compile  |  21m 32s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |  18m  5s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   3m  5s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   7m 21s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 17s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 45s |  hadoop-common in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 50s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 40s |  hadoop-hdfs-httpfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 41s |  hadoop-hdfs-rbf in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 41s |  hadoop-yarn-server-nodemanager in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 41s |  hadoop-yarn-server-resourcemanager in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 39s |  hadoop-mapreduce-client-hs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   5m 35s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   1m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  11m 43s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 27s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 27s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |  20m 27s |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 58s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  17m 58s |  the patch passed  |
| -0 :warning: |  checkstyle  |   3m 32s |  root: The patch generated 7 new + 899 unchanged - 8 fixed = 906 total (was 907)  |
| -1 :x: |  mvnsite  |   1m  3s |  hadoop-hdfs in the patch failed.  |
| -1 :x: |  mvnsite  |   0m 33s |  hadoop-hdfs-httpfs in the patch failed.  |
| -1 :x: |  mvnsite  |   0m 43s |  hadoop-hdfs-rbf in the patch failed.  |
| -1 :x: |  mvnsite  |   0m 40s |  hadoop-yarn-server-nodemanager in the patch failed.  |
| -1 :x: |  mvnsite  |   0m 44s |  hadoop-yarn-server-resourcemanager in the patch failed.  |
| -1 :x: |  mvnsite  |   0m 41s |  hadoop-mapreduce-client-hs in the patch failed.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m 23s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 43s |  hadoop-common in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 46s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 36s |  hadoop-hdfs-httpfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 35s |  hadoop-hdfs-rbf in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 38s |  hadoop-yarn-server-nodemanager in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 38s |  hadoop-yarn-server-resourcemanager in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| -1 :x: |  javadoc  |   0m 33s |  hadoop-mapreduce-client-hs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   5m 32s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| -1 :x: |  findbugs  |   1m  3s |  hadoop-hdfs in the patch failed.  |
| -1 :x: |  findbugs  |   0m 39s |  hadoop-hdfs-httpfs in the patch failed.  |
| -1 :x: |  findbugs  |   0m 38s |  hadoop-hdfs-rbf in the patch failed.  |
| -1 :x: |  findbugs  |   0m 42s |  hadoop-yarn-server-nodemanager in the patch failed.  |
| -1 :x: |  findbugs  |   0m 50s |  hadoop-yarn-server-resourcemanager in the patch failed.  |
| -1 :x: |  findbugs  |   0m 45s |  hadoop-mapreduce-client-hs in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   9m 23s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  |   0m 56s |  hadoop-hdfs in the patch failed.  |
| -1 :x: |  unit  |   0m 34s |  hadoop-hdfs-httpfs in the patch failed.  |
| -1 :x: |  unit  |   0m 40s |  hadoop-hdfs-rbf in the patch failed.  |
| -1 :x: |  unit  |   0m 44s |  hadoop-yarn-server-nodemanager in the patch failed.  |
| -1 :x: |  unit  |   0m 42s |  hadoop-yarn-server-resourcemanager in the patch failed.  |
| -1 :x: |  unit  |   0m 39s |  hadoop-mapreduce-client-hs in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 50s |  The patch does not generate ASF License warnings.  |
|  |   | 212m 27s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.security.TestRuleBasedLdapGroupsMapping |
|   | hadoop.security.TestGroupsCaching |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2085 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 21e84f1d09ef 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / bed0a3a3740 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-httpfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-rbf-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-nodemanager-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-resourcemanager-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/branch-javadoc-hadoop-mapreduce-project_hadoop-mapreduce-client_hadoop-mapreduce-client-hs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/diff-checkstyle-root.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-mvnsite-hadoop-hdfs-project_hadoop-hdfs.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-mvnsite-hadoop-hdfs-project_hadoop-hdfs-httpfs.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-mvnsite-hadoop-hdfs-project_hadoop-hdfs-rbf.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-mvnsite-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-nodemanager.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-mvnsite-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-resourcemanager.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-mvnsite-hadoop-mapreduce-project_hadoop-mapreduce-client_hadoop-mapreduce-client-hs.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-httpfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-rbf-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-nodemanager-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-resourcemanager-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-javadoc-hadoop-mapreduce-project_hadoop-mapreduce-client_hadoop-mapreduce-client-hs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-findbugs-hadoop-hdfs-project_hadoop-hdfs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-findbugs-hadoop-hdfs-project_hadoop-hdfs-httpfs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-findbugs-hadoop-hdfs-project_hadoop-hdfs-rbf.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-findbugs-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-nodemanager.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-findbugs-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-resourcemanager.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-findbugs-hadoop-mapreduce-project_hadoop-mapreduce-client_hadoop-mapreduce-client-hs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs-httpfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs-rbf.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-nodemanager.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-yarn-project_hadoop-yarn_hadoop-yarn-server_hadoop-yarn-server-resourcemanager.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/artifact/out/patch-unit-hadoop-mapreduce-project_hadoop-mapreduce-client_hadoop-mapreduce-client-hs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/testReport/ |
| Max. process+thread count | 1838 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-httpfs hadoop-hdfs-project/hadoop-hdfs-rbf hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2085/5/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
@jojochuang can you help take a look at this change? Thanks

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/LdapGroupsMapping.java [hunk 10]
```diff
@@ -320,21 +321,21 @@
   private Iterator<BindUserInfo> bindUsers;
   private BindUserInfo currentBindUser;
 
-  private String userbaseDN;
+  private volatile String userbaseDN;
```
Comment (jojochuang) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/LdapGroupsMapping.java:324]: Why do we use volatile variables here?

Comment (jojochuang) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/LdapGroupsMapping.java:324]: Just realized it's there to make findbugs happy.

**Comment by xiaoyuyao** [comment]
Thanks @Hexiaoqiao and @jojochuang  for the review. I've merged the PR to trunk.
